### PR TITLE
Decoder.

### PR DESCRIPTION
**Pull request by timav566**
Decoder and training and validation steps were implemented.

P.S. run "train_model_test" to see the results.

### CODE REVIEW

File: running/util.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import numpy as np
+import torch
+
+
+# Based on https://github.com/DongjunLee/transformer-tensorflow/blob/master/transformer/attention.py
+def positional_encoding(dim, sentence_length, dtype=torch.float32):
```
[the remark this answers is not in the document]
Comment (egor-bogomolov) [reply running/util.py:6]: It is better not to use constant tensors in your models without wrapping them into Parameter or registering them as buffers. Things to consider are:
1. Should the tensor update during training (does it require gradient)? -- `Parameter(...)`
2. Should user see it alongside model parameters? -- Rare case when you really want it to be among the parameters but do not need to update it -- `Parameter(..., requires_grad=False)`
3. Ok, not among the parameters, but belongs to the models' state and should be together with weights when dumping the model into a checkpoint file -- `self.register_buffer("tensor_name", tensor_name)`
4. Well, it's something big that I don't want to store in the checkpoint, but it still should be transferred to GPU/CPU/multiple GPUs in distributed training/... -- `self.register_buffer("tensor_name", tensor_name, persistent=False)`

Also Autograd will have more information about it and can work with it more efficiently.


In this specific case when you have some non-trivial code to compute tensor values + know how it will be applied, it is better to turn positional encoding into a separate module while registering the values as buffer (see [here](https://pytorch.org/tutorials/beginner/transformer_tutorial.html) for example).

[the remark this answers is not in the document]
Comment (egor-bogomolov) [reply running/util.py:6]: [A really good thread explaining different use-cases](https://discuss.pytorch.org/t/what-is-the-difference-between-register-buffer-and-register-parameter-of-nn-module/32723)

File: running/util.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+            for pos in range(sentence_length)
+            for i in range(dim)
+        ]
+    )
+    encoded_vec[::2] = np.sin(encoded_vec[::2])
+    encoded_vec[1::2] = np.cos(encoded_vec[1::2])
+    return torch.tensor(encoded_vec.reshape([sentence_length, dim]), dtype=dtype)
+
+
+def prefix_sum(arr):
```
Comment (egor-bogomolov) [inline running/util.py:19]: [Use built-ins](https://pytorch.org/docs/stable/generated/torch.cumsum.html)

File: running/wraped_model.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        subtoken_embeddings = torch.index_select(self.embedding, 0, flat_tokens)
+        subtoken_embeddings = torch.reshape(subtoken_embeddings, original_shape)
+        subtoken_embeddings *= torch.unsqueeze(torch.clamp(tokens, 0, 1), -1).cuda()
+        states = torch.mean(subtoken_embeddings, 2)
+        # have to understand why the following line is needed
+        # states += self.pos_enc[:states.shape[1]]
+        predictions = torch.transpose(self.prediction(self.model(states)[0]), 1, 2)
+        return predictions
+
+    def training_step(self, batch, batch_idx):
```
Comment (egor-bogomolov) [inline running/wraped_model.py:58]: Missing type annotations

File: running/wraped_model.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        tokens, edges, error_loc, repair_targets, repair_candidates = batch
+        token_mask = torch.clamp(torch.sum(tokens, -1), 0, 1)
+        pointer_preds = self(tokens, token_mask, edges)
+        ls, acs = self.get_loss(
+            pointer_preds, token_mask, error_loc, repair_targets, repair_candidates
+        )
+        loss = sum(ls)
+        return loss
+
+    def validation_step(self, batch, batch_idx):
```
Comment (egor-bogomolov) [inline running/wraped_model.py:69]: Missing type annotations

File: running/wraped_model.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        self.log("val_bug_loc_acc", acs[1], prog_bar=True)
+        self.log("val_target_loc_acc", acs[2], prog_bar=True)
+        self.log("val_joint_acc", acs[3], prog_bar=True)
+        return loss
+
+    def test_step(self, batch, batch_idx):
+        # Here we just reuse the validation_step for testing
+        return self.validation_step(batch, batch_idx)
+
+    def configure_optimizers(self):
```
Comment (egor-bogomolov) [inline running/wraped_model.py:89]: Missing type annotations

File: running/wraped_model.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        return torch.optim.Adam(
+            self.parameters(), lr=self.training_config["learning_rate"]
+        )
+
+    # probably there are lots of bugs here right now...
+    def get_loss(
```
Comment (egor-bogomolov) [inline running/wraped_model.py:95]: Missing type annotations

File: running/util.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    for a in arr:
+        res.append(res[-1] + a)
+    return res
+
+
+def sparse_categorical_accuracy(y_true, y_pred):
+    return torch.eq(y_true, torch.argmax(y_pred, -1)).long()
+
+
+def sparse_softmax_cross_entropy_with_logits(error_locations, loc_predictions):
```
Comment (egor-bogomolov) [inline running/util.py:30]: Missing type annotations

File: running/util.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+
+def prefix_sum(arr):
+    res = [0]
+    for a in arr:
+        res.append(res[-1] + a)
+    return res
+
+
+def sparse_categorical_accuracy(y_true, y_pred):
```
Comment (egor-bogomolov) [inline running/util.py:26]: Missing type annotations

File: running/wraped_model.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+            raise ValueError("Unknown model component provided:", inner_model)
+
+    def forward(self, tokens, token_mask, edges):
+        original_shape = list(
+            np.append(np.array(tokens.shape), self.model_config["base"]["hidden_dim"])
+        )
+        flat_tokens = tokens.type(torch.LongTensor).flatten().cuda()
+        subtoken_embeddings = torch.index_select(self.embedding, 0, flat_tokens)
+        subtoken_embeddings = torch.reshape(subtoken_embeddings, original_shape)
+        subtoken_embeddings *= torch.unsqueeze(torch.clamp(tokens, 0, 1), -1).cuda()
```
Comment (egor-bogomolov) [inline running/wraped_model.py:51]: If you use built-in Embedding, you won't need any of it most likely

File: running/wraped_model.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+        states = torch.mean(subtoken_embeddings, 2)
+        # have to understand why the following line is needed
+        # states += self.pos_enc[:states.shape[1]]
```
Comment (egor-bogomolov) [inline running/wraped_model.py:54]: For RNN you don't need the positional encoding at all. It is needed for Transformers as they don't have a clue about positions of tokens in the sequence otherwise

File: running/wraped_model.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+        predictions = torch.transpose(self.prediction(self.model(states)[0]), 1, 2)
+        return predictions
+
+    def training_step(self, batch, batch_idx):
+        torch.cuda.empty_cache()
```
Comment (egor-bogomolov) [inline running/wraped_model.py:59]: Why is it needed?

File: running/wraped_model.py [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+        ls, acs = self.get_loss(
+            pointer_preds, token_mask, error_loc, repair_targets, repair_candidates
+        )
+        loss = sum(ls)
+        self.log("loc_loss", ls[0], prog_bar=True)
+        self.log("target_loss", ls[1], prog_bar=True)
+        self.log("val_no_bug_pred_acc", acs[0], prog_bar=True)
+        self.log("val_bug_loc_acc", acs[1], prog_bar=True)
+        self.log("val_target_loc_acc", acs[2], prog_bar=True)
+        self.log("val_joint_acc", acs[3], prog_bar=True)
```
Comment (egor-bogomolov) [inline running/wraped_model.py:82]: There is `log_dict` method in lightning. Most likely it would be more convenient to return a dictionary from the `self.get_loss` function rather than parsing it this way

File: running/wraped_model.py [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+        loss = sum(ls)
+        return loss
+
+    def validation_step(self, batch, batch_idx):
+        tokens, edges, error_loc, repair_targets, repair_candidates = batch
+        token_mask = torch.clamp(torch.sum(tokens, -1), 0, 1)
+        pointer_preds = self(tokens, token_mask, edges)
+        ls, acs = self.get_loss(
+            pointer_preds, token_mask, error_loc, repair_targets, repair_candidates
+        )
```
Comment (egor-bogomolov) [inline running/wraped_model.py:75]: Code duplicates `training_step`. Consider using the same code for training, validation, and testing steps (for examples see [shared_step here](https://github.com/JetBrains-Research/code2seq/blob/master/code2seq/model/code2seq.py))

File: running/wraped_model.py [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+        self.log("target_loss", ls[1], prog_bar=True)
+        self.log("val_no_bug_pred_acc", acs[0], prog_bar=True)
+        self.log("val_bug_loc_acc", acs[1], prog_bar=True)
+        self.log("val_target_loc_acc", acs[2], prog_bar=True)
+        self.log("val_joint_acc", acs[3], prog_bar=True)
+        return loss
+
+    def test_step(self, batch, batch_idx):
+        # Here we just reuse the validation_step for testing
+        return self.validation_step(batch, batch_idx)
```
Comment (egor-bogomolov) [inline running/wraped_model.py:87]: All the metrics will be logged with `val` prefix instead of `test`. See the comment about shared step above.

File: running/wraped_model.py (hunk 12 above)
Comment (egor-bogomolov) [reply running/wraped_model.py:82]: Also, does `prog_bar=True` behave correctly when you log so much stuff one by one?

File: running/wraped_model.py [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+        loc_predictions = predictions[:, 0]
+        loc_loss = sparse_softmax_cross_entropy_with_logits(
+            error_locations, loc_predictions
+        )
+        loc_loss = torch.mean(loc_loss)
+        loc_accs = sparse_categorical_accuracy(error_locations, loc_predictions)
+        no_bug_pred_acc = torch.sum((1 - is_buggy) * loc_accs) / (
+            1e-9 + torch.sum(1 - is_buggy)
+        )
+        # I added this because in case torch.sum(1 - is_buggy) == 0 was calculated wrong
```
Comment (egor-bogomolov) [inline running/wraped_model.py:117]: Why not fix it right away? :slightly_smiling_face:

File: running/wraped_model.py [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+        is_buggy = torch.clamp(error_locations, 0, 1).float()
+        loc_predictions = predictions[:, 0]
+        loc_loss = sparse_softmax_cross_entropy_with_logits(
+            error_locations, loc_predictions
+        )
+        loc_loss = torch.mean(loc_loss)
+        loc_accs = sparse_categorical_accuracy(error_locations, loc_predictions)
+        no_bug_pred_acc = torch.sum((1 - is_buggy) * loc_accs) / (
+            1e-9 + torch.sum(1 - is_buggy)
+        )
```
Comment (egor-bogomolov) [inline running/wraped_model.py:116]: You mix computations of metrics and loss here. It would be better to have separate methods and call only the one that is needed at the moment.

File: running/wraped_model.py [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+        # I added this because in case torch.sum(1 - is_buggy) == 0 was calculated wrong
+        if torch.sum(1 - is_buggy) == 0:
+            no_bug_pred_acc = 1
+        bug_loc_acc = torch.sum(is_buggy * loc_accs) / (1e-9 + torch.sum(is_buggy))
+        pointer_logits = predictions[:, 1]
+
+        # maybe there is an appropriate function in pytorch for that
+        candidate_mask = np.zeros(pointer_logits.size())
```
Comment (egor-bogomolov) [inline running/wraped_model.py:124]: See [zeros_like](https://pytorch.org/docs/stable/generated/torch.zeros_like.html) (also exists in numpy)

File: running/wraped_model.py [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+        for e in repair_candidates:
+            candidate_mask[e[0]][e[1]] = 1
+        candidate_mask = torch.tensor(candidate_mask).cuda()
```
Comment (egor-bogomolov) [inline running/wraped_model.py:127]: `cuda()` calls will strike back if you want to use non-cuda device. Better use `device=self.device` from Lightning

File: test/train_model_test.py [hunk 19]
```diff
@@ -0,0 +1,26 @@
+import pytorch_lightning as pl
+from running.wraped_model import VarMisuseLayer
+import yaml
+from data_processing import vocabulary, graph_data_loader
+
+data_path = "/home/timav/jb/graph-code-nets/data"
+config_path = "/home/timav/jb/graph-code-nets/config.yml"
+vocabulary_path = "/home/timav/jb/graph-code-nets/vocab.txt"
```
Comment (egor-bogomolov) [inline test/train_model_test.py:8]: Pushing your local paths to the repository is a bad practice, this code won't run for anyone except you. Please, load the paths from config rather than hard-code and make them relative.

File: test/train_model_test.py [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+
+config = yaml.safe_load(open(config_path))
```
Comment (egor-bogomolov) [inline test/train_model_test.py:10]: Config path should be passed as a parameter when calling the script

File: test/util_test.py [hunk 21]
```diff
@@ -0,0 +1,23 @@
+import torch
```
Comment (egor-bogomolov) [inline test/util_test.py:1]: Consider turning it into real tests with pytest or unittest and add the respective line to the CI config

**Review by egor-bogomolov — CHANGES_REQUESTED** [review]
Good job! There are no type annotations at the moment, please add them (both for arguments and for return values). Also, there are several issues with the model implementation that should be fixed before merging.

I did not check notebooks for now, will do so after fixes to the scripts.

File: models/rnn.py [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+
+class RNN(pl.LightningModule):
+    def __init__(self, model_config, shared_embedding=None, vocab_dim=None):
+        super().__init__()
+        self.hidden_dim = model_config["hidden_dim"]
+        self.num_layers = model_config["num_layers"]
+        self.dropout_rate = model_config["dropout_rate"]
+        self.rnn = torch.nn.GRU(
+            input_size=self.hidden_dim,
+            hidden_size=self.hidden_dim // 2,
```
[the remark this answers is not in the document]
Comment (timav566) [reply models/rnn.py:13]: because of _bidirectional=True_, so the output size will be exactly _hidden_dim_

File: running/util.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (timav566) [reply running/util.py:6]: I decided to temporarily remove positional encoding, until I don't have a transformer model right now, but I'll work on it later. 
Btw very interesting thread about register buffer and parameter :)

File: running/wraped_model.py (hunk 12 above)
Comment (timav566) [reply running/wraped_model.py:82]: yes, it behaves correctlty but I agree with the point that it's better to use dict here.
